### ISSUE TITLE
The compiler incorrectly optimizes `(1, 1) as (int, int x)?` to `null`

### ISSUE THREAD

**Issue by VladimirReshetnikov** [dotnet/roslyn#17963]
**Version Used**: 

Microsoft Visual Studio Enterprise 2017
Version 15.0.26228.9 D15RTWSVC
Microsoft .NET Framework
Version 4.6.01586

**Steps to Reproduce**:

Compile and run:

```
using System;

class P
{
    static void Main()
    {
        var x = (1, 1) as (int, int x)?;
        Console.WriteLine(x == null);

        var y = (1, new object()) as (int, dynamic)?;
        Console.WriteLine(y == null);
    }
}
```

**Expected Behavior**: The program compiles successfully and prints: False, False

**Actual Behavior**:
warning CS0458: The result of the expression is always 'null' of type '(int, int x)?'
warning CS0458: The result of the expression is always 'null' of type '(int, dynamic)?'
The program prints: True, True

**Comment by jcouv** [comment]
Thanks for the bug reports!
Interestingly, the behavior on the first case changes when no tuple element name is used in the `as`.
`var x = (1, 1) as (int, int x)?;`

**Comment by jcouv** [comment]
FYI @VSadov

**Comment by VSadov** [comment]
Bug in constant folding.
Should fix before this is a compat burden.

**Comment by jcouv** [comment]
From quick check with @gafter, we actually expect `(1, 2) as (int, int)?` (and therefore `(1, 2) as (int, int b)?`) to be non-null.

**Comment by jcouv** [comment]
Thanks for reporting this issue. The fix (#19697) will ship in Visual Studio 2017 version 15.3.
